File: python_backend/archive/msc_scraper.py
import requests
import json
import time
import os
from DrissionPage import ChromiumOptions, ChromiumPage
from db_manager import get_db_connection

def update_msc_tokens_in_db(access_token):
    import json
    token_payload = json.dumps({"MSC_ACCESS_TOKEN": access_token})
    
    upsert_query = """
        INSERT INTO site_tokens (site_name, token_data, last_updated)
        VALUES (%s, %s, CURRENT_TIMESTAMP)
        ON CONFLICT (site_name)
        DO UPDATE SET token_data = EXCLUDED.token_data, last_updated = CURRENT_TIMESTAMP;
    """
    
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute(upsert_query, ('MSC', token_payload))
    conn.commit()
    cursor.close()
    conn.close()

def get_msc_token_from_db():
    import json
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT token_data FROM site_tokens WHERE site_name = 'MSC';")
    row = cursor.fetchone()
    cursor.close()
    conn.close()
    if row:
        token_data = row[0]
        if isinstance(token_data, str):
            token_data = json.loads(token_data)
        return token_data.get('MSC_ACCESS_TOKEN')
    return None

def fetch_msc_tokens_background():
    """
    Uses DrissionPage in headless mode to navigate to MSC, login, and capture the bearer token.
    """
    print("Starting MSC token capture in background...")
    co = ChromiumOptions()
    
    is_headless = os.getenv('SCRAPER_HEADLESS', 'true').lower() == 'true'
    co.headless(is_headless)
    
    co.set_user_data_path('./msc_drission_user_data')
    co.auto_port()
    
    # Prevent timeouts on slower systems
    co.set_timeouts(base=20, page_load=30)
    
    page = ChromiumPage(co)
    
    try:
        # Start directly at the myMSC portal where the login form is embedded
        print("Navigating to myMSC Login Portal...")
        page.get("https://www.mymsc.com/myMSC/")
        
        # Click cookie banner if it exists ("Accept All" button shown in screenshot)
        cookie_btn = page.ele('text:Accept All', timeout=5) or page.ele('@id=onetrust-accept-btn-handler')
        if cookie_btn:
            print("Accepting cookies...")
            cookie_btn.click()
            time.sleep(1)
            
        print("Looking for MSC email field...")
        email_field = page.ele('@name=signInName', timeout=10) or page.ele('#signInName') or page.ele('@type=email') or page.ele('@name=logonIdentifier')
        if email_field:
            print("Entering email...")
            email_field.input('alejandro.delcarpio@primeteam.com.mx')
            
            # The Next button shown in the screenshot
            next_btn = page.ele('@id=next') or page.ele('text:Next') or page.ele('text:Sign in') or page.ele('@type=submit')
            if next_btn:
                print("Clicking Next...")
                next_btn.click()
                time.sleep(3)
                
            print("Looking for password field...")
            pass_field = page.ele('@name=password', timeout=10) or page.ele('#password') or page.ele('@type=password')
            if pass_field:
                print("Entering password...")
                pass_field.input('Adc19770123$')
                
                # The final submit button after password
                submit_btn = page.ele('@id=next') or page.ele('text:Sign in') or page.ele('text:Next') or page.ele('@type=submit') or page.ele('@id=continue')
                if submit_btn:
                    print("Submitting login...")
                    # Listen to ALL requests to ensure we don't miss the token
                    page.listen.start()
                    submit_btn.click()
                    
                    # Wait for Microsoft B2C login redirects to finish (this can take up to 25 seconds on slow networks)
                    print("Waiting for Microsoft B2C redirects to finish...")
                    time.sleep(25)
                    
                    # Force a navigation to the quote dashboard to guarantee an API call is made with the Bearer token
                    print("Navigating to Quotes page to trigger API call...")
                    page.get("https://www.mymsc.com/myMSC/Quotes/Request")
                    
                    # Wait for the API call that has the Bearer token
                    found_token = None
                    print("Listening for MSC Bearer token...")
                    for packet in page.listen.steps(timeout=30):
                        req_headers = packet.request.headers
                        auth_header = req_headers.get('Authorization') or req_headers.get('authorization')
                        if auth_header and 'Bearer' in auth_header:
                            found_token = auth_header.split('Bearer ')[1]
                            break
                    
                    if found_token:
                        print("Successfully intercepted MSC token!")
                        update_msc_tokens_in_db(found_token)
                    else:
                        print("Failed to intercept MSC token. Check if login was successful.")
        else:
            print("Could not find the MSC login fields. Are we on the right page?")
            
    except Exception as e:
        print(f"Error capturing MSC tokens: {e}")
    finally:
        page.quit()

def fetch_msc_quotes(origin, destination, equipment, commodity="FAK"):
    """
    Uses python requests with the DB token to fetch MSC rates.
    """
    token = get_msc_token_from_db()
    if not token:
        return {"success": False, "error": "MSC Token not found. Scraper may still be initializing."}
        
    print(f"Fetching MSC rates for {origin} -> {destination} using Live Requests API!")
    
    url = 'https://services.mymsc.com/quote/graphql'
    
    headers = {
        'accept': '*/*',
        'accept-language': 'en-US,en;q=0.9',
        'authorization': f'Bearer {token}',
        'content-type': 'application/json',
        'mymsc-user-email': 'alejandro.delcarpio@primeteam.com.mx',
        'origin': 'https://www.mymsc.com',
        'referer': 'https://www.mymsc.com/',
        'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Safari/537.36'
    }
    
    # We are temporarily hardcoding the Shanghai -> Barcelona IDs until UNLOCODE lookup is implemented
    # Shanghai = 7759, Barcelona = 8834
    
    payload = {
        "query": "\n    query InstantQuoteSearchV5($input: RateCardSearchCriteriaInput!) {\n  searchRateCardsV5(request: $input) {\n    shippingWindowBasedGroups {\n      rateCards {\n        myMscId\n        sizeType\n        total\n        subTotal\n        currency\n        scheduleInformation {\n          oceanTransitDays\n          departureDate\n          arrivalDate\n          serviceName\n        }\n      }\n    }\n    vesselBasedGroups {\n      rateCards {\n        myMscId\n        sizeType\n        total\n        subTotal\n        currency\n        scheduleInformation {\n          oceanTransitDays\n          departureDate\n          arrivalDate\n          serviceName\n        }\n      }\n    }\n  }\n}\n    ",
        "variables": {
            "input": {
                "originId": 7759,
                "isOriginAPort": True,
                "destinationId": 8834,
                "isDestinationAPort": True,
                "originTransportationMode": "",
                "destinationTransportationMode": "",
                "originZipcode": "",
                "destinationZipcode": "",
                "equipmentFilter": [
                    {
                        "equipmentType": "20DV" if "20" in equipment else "40HC",
                        "weightValue": 18000
                    }
                ],
                "weightUnit": "Kgs",
                "commodityGroupCode": "",
                "temperature": None,
                "temperatureUnit": None,
                "cargoValue": None,
                "cargoDestinationCountryId": 230,
                "cargoOriginCountryId": 195,
                "commodityGroupDescription": ""
            }
        },
        "operationName": "InstantQuoteSearchV5"
    }

    try:
        response = requests.post(url, headers=headers, json=payload, timeout=20)
        
        if response.status_code == 401:
            return {"success": False, "error": "MSC Token expired or invalid (401). Background scraper needs to refresh."}
            
        response.raise_for_status()
        res_json = response.json()
        
        # Parse GraphQL response
        data = res_json.get('data', {}).get('searchRateCardsV5', {})
        if not data:
            return {"success": False, "error": "Invalid GraphQL response format from MSC"}
            
        groups = data.get('shippingWindowBasedGroups', []) + data.get('vesselBasedGroups', [])
        
        formatted_quotes = []
        for group in groups:
            rate_cards = group.get('rateCards', [])
            for rc in rate_cards:
                sched = rc.get('scheduleInformation', {})
                formatted_quotes.append({
                    "scheduleId": rc.get('myMscId', 'MSC-UNKNOWN'),
                    "departureDate": sched.get('departureDate', ''),
                    "arrivalDate": sched.get('arrivalDate', ''),
                    "transitTime": sched.get('oceanTransitDays', 0),
                    "vesselName": sched.get('serviceName', 'MSC Vessel'),
                    "equipmentName": rc.get('sizeType', ''),
                    "co2": 0,
                    "prices": [
                        {
                            "equipmentType": rc.get('sizeType', ''),
                            "oceanFreight": rc.get('subTotal', 0),
                            "totalCharge": rc.get('total', 0),
                            "currency": rc.get('currency', 'USD')
                        }
                    ]
                })
                
        return {
            "success": True,
            "data": formatted_quotes
        }
    except Exception as e:
        print(f"MSC API Error: {e}")
        return {"success": False, "error": str(e)}

File: xportplus_backend/src/scraper/MscScraper.js
// const BaseScraper = require('./BaseScraper');
// const { saveToken } = require('../../config/db');

// class MscScraper extends BaseScraper {
//   constructor() {
//     super('MSC');
//     this.email = 'alejandro.delcarpio@primeteam.com.mx';
//     this.password = 'Adc19770123$';
//   }

//   async scrapeTokens() {
//     try {
//       console.log('Starting MSC scraper...');
//       await this.initBrowser();

//       let bearerToken = '';

//       await this.page.setRequestInterception(true);
//       this.page.on('request', (request) => {
//         const headers = request.headers();

//         if (request.url().includes('services.mymsc.com/quote/graphql') || request.url().includes('services.mymsc.com')) {
//           if (headers['authorization'] && headers['authorization'].startsWith('Bearer ')) {
//             bearerToken = headers['authorization'].split(' ')[1];
//           }
//         }
//         request.continue();
//       });

//       await this.page.goto('https://www.mymsc.com/', { waitUntil: 'networkidle2' });

//       // Handle cookies
//       try {
//         await this.page.waitForSelector('#onetrust-accept-btn-handler', { timeout: 3000 });
//         await this.page.click('#onetrust-accept-btn-handler');
//       } catch (e) {}

//       // Login
//       const isLoginVisible = await this.page.$('input[type="email"]');
//       if (isLoginVisible) {
//         console.log('Logging into MSC...');
//         await this.page.type('input[type="email"]', this.email);
//         await this.page.keyboard.press('Enter');

//         try {
//           await new Promise(r => setTimeout(r, 5000)); // hard wait for page transition
//           await this.page.waitForSelector('input[type="password"]', { timeout: 10000 });
//           await this.page.type('input[type="password"]', this.password);
//           await this.page.keyboard.press('Enter');

//           await this.page.waitForNavigation({ waitUntil: 'networkidle2', timeout: 15000 }).catch(() => {});
//         } catch (err) {
//           console.log('Could not find password field or navigation failed:', err.message);
//         }
//       }

//       console.log('Waiting for tokens to be intercepted...');
//       await new Promise(r => setTimeout(r, 10000)); 

//       // Fallback: extract JWT token from anywhere in localStorage or sessionStorage
//       if (!bearerToken) {
//         bearerToken = await this.page.evaluate(() => {
//           let token = null;
//           // Look in localStorage
//           for (let i = 0; i < localStorage.length; i++) {
//             const key = localStorage.key(i);
//             const val = localStorage.getItem(key);
//             if (val && val.includes('eyJ') && val.length > 500) {
//               token = val;
//             }
//           }
//           // Look in sessionStorage
//           for (let i = 0; i < sessionStorage.length; i++) {
//             const key = sessionStorage.key(i);
//             const val = sessionStorage.getItem(key);
//             if (val && val.includes('eyJ') && val.length > 500) {
//               token = val;
//             }
//           }

//           // sometimes it's wrapped in JSON
//           if (token && token.startsWith('{')) {
//              try {
//                 const parsed = JSON.parse(token);
//                 if (parsed.secret) token = parsed.secret;
//                 else if (parsed.credential) token = parsed.credential;
//              } catch(e) {}
//           }
//           return token;
//         });
//       }

//       if (bearerToken) {
//         const tokenData = {
//           MSC_BEARER_TOKEN: bearerToken
//         };
//         await saveToken(this.siteName, tokenData);
//         console.log('MSC tokens saved to database.');
//       } else {
//         console.log('Failed to intercept MSC tokens. Taking screenshot...');
//         await this.page.screenshot({ path: 'msc_failed.png' });
//       }

//     } catch (error) {
//       console.error('Error during MSC scraping:', error);
//     } finally {
//       await this.closeBrowser();
//     }
//   }
// }

// module.exports = MscScraper;


// const BaseScraper = require('./BaseScraper');
// const { saveToken } = require('../config/db');

// class MscScraper extends BaseScraper {
//   constructor() {
//     super('MSC');
//     this.email = process.env.MSC_EMAIL;
//     this.password = process.env.MSC_PASSWORD;
//   }

//   async scrapeTokens() {
//     if (!this.email || !this.password) {
//       console.error('MSC_EMAIL / MSC_PASSWORD are not set — skipping MSC scrape.');
//       return;
//     }

//     try {
//       console.log('Starting MSC scraper...');
//       await this.initBrowser();

//       let bearerToken = '';

//       await this.page.setRequestInterception(true);
//       this.page.on('request', (request) => {
//         const headers = request.headers();
//         if (request.url().includes('services.mymsc.com')) {
//           if (headers['authorization'] && headers['authorization'].startsWith('Bearer ')) {
//             bearerToken = headers['authorization'].split(' ')[1];
//           }
//         }
//         request.continue();
//       });

//       await this.page.goto('https://www.mymsc.com/', { waitUntil: 'networkidle2' });

//       try {
//         await this.page.waitForSelector('#onetrust-accept-btn-handler', { timeout: 3000 });
//         await this.page.click('#onetrust-accept-btn-handler');
//       } catch (e) { }

//       const isLoginVisible = await this.page.$('input[type="email"]');
//       if (isLoginVisible) {
//         await this.page.type('input[type="email"]', this.email);
//         await this.page.keyboard.press('Enter');
//         try {
//           await new Promise(r => setTimeout(r, 5000));
//           await this.page.waitForSelector('input[type="password"]', { timeout: 10000 });
//           await this.page.type('input[type="password"]', this.password);
//           await this.page.keyboard.press('Enter');
//           await this.page.waitForNavigation({ waitUntil: 'networkidle2', timeout: 15000 }).catch(() => { });
//         } catch (err) {
//           console.log('Could not find password field or navigation failed:', err.message);
//         }
//       }

//       await new Promise(r => setTimeout(r, 10000));

//       if (!bearerToken) {
//         bearerToken = await this.page.evaluate(() => {
//           let token = null;
//           for (let i = 0; i < localStorage.length; i++) {
//             const val = localStorage.getItem(localStorage.key(i));
//             if (val && val.includes('eyJ') && val.length > 500) token = val;
//           }
//           for (let i = 0; i < sessionStorage.length; i++) {
//             const val = sessionStorage.getItem(sessionStorage.key(i));
//             if (val && val.includes('eyJ') && val.length > 500) token = val;
//           }
//           if (token && token.startsWith('{')) {
//             try {
//               const parsed = JSON.parse(token);
//               if (parsed.secret) token = parsed.secret;
//               else if (parsed.credential) token = parsed.credential;
//             } catch (e) { }
//           }
//           return token;
//         });
//       }

//       if (bearerToken) {
//         await saveToken(this.siteName, { MSC_BEARER_TOKEN: bearerToken });
//         console.log('MSC tokens saved to database.');
//       } else {
//         console.log('Failed to intercept MSC tokens. Taking screenshot...');
//         await this.page.screenshot({ path: 'msc_failed.png' });
//       }
//     } catch (error) {
//       console.error('Error during MSC scraping:', error);
//     } finally {
//       await this.closeBrowser();
//     }
//   }
// }

// module.exports = MscScraper;


const BaseScraper = require('./BaseScraper');
const { saveToken } = require('../config/db');

class MscScraper extends BaseScraper {
  constructor() {
    super('MSC');
    this.email = process.env.MSC_EMAIL;
    this.password = process.env.MSC_PASSWORD;
  }

  async scrapeTokens() {
    if (!this.email || !this.password) {
      console.error('MSC_EMAIL / MSC_PASSWORD are not set — skipping MSC scrape.');
      return;
    }

    try {
      console.log('Starting MSC scraper...');
      await this.initBrowser();

      let bearerToken = '';

      await this.page.setRequestInterception(true);
      this.page.on('request', (request) => {
        const headers = request.headers();
        if (request.url().includes('services.mymsc.com')) {
          if (headers['authorization'] && headers['authorization'].startsWith('Bearer ')) {
            bearerToken = headers['authorization'].split(' ')[1];
          }
        }
        request.continue();
      });

      await this.page.goto('https://www.mymsc.com/', { waitUntil: 'networkidle2' });

      try {
        await this.page.waitForSelector('#onetrust-accept-btn-handler', { timeout: 3000 });
        await this.page.click('#onetrust-accept-btn-handler');
      } catch (e) { }

      const isLoginVisible = await this.page.$('input[type="email"]');
      if (isLoginVisible) {
        await this.page.type('input[type="email"]', this.email);
        await this.page.keyboard.press('Enter');
        try {
          await new Promise(r => setTimeout(r, 5000));
          await this.page.waitForSelector('input[type="password"]', { timeout: 10000 });
          await this.page.type('input[type="password"]', this.password);
          await this.page.keyboard.press('Enter');
          await this.page.waitForNavigation({ waitUntil: 'networkidle2', timeout: 15000 }).catch(() => { });
        } catch (err) {
          console.log('Could not find password field or navigation failed:', err.message);
        }
      }

      await new Promise(r => setTimeout(r, 10000));

      if (!bearerToken) {
        bearerToken = await this.page.evaluate(() => {
          let token = null;
          for (let i = 0; i < localStorage.length; i++) {
            const val = localStorage.getItem(localStorage.key(i));
            if (val && val.includes('eyJ') && val.length > 500) token = val;
          }
          for (let i = 0; i < sessionStorage.length; i++) {
            const val = sessionStorage.getItem(sessionStorage.key(i));
            if (val && val.includes('eyJ') && val.length > 500) token = val;
          }
          if (token && token.startsWith('{')) {
            try {
              const parsed = JSON.parse(token);
              if (parsed.secret) token = parsed.secret;
              else if (parsed.credential) token = parsed.credential;
            } catch (e) { }
          }
          return token;
        });
      }

      if (bearerToken) {
        await saveToken(this.siteName, { MSC_BEARER_TOKEN: bearerToken });
        console.log('MSC tokens saved to database.');
      } else {
        console.log('Failed to intercept MSC tokens. Taking screenshot...');
        await this.page.screenshot({ path: 'msc_failed.png' });
      }
    } catch (error) {
      console.error('Error during MSC scraping:', error);
    } finally {
      await this.closeBrowser();
    }
  }
}

module.exports = MscScraper;